apollo-federation/src/composition/satisfiability/validation_state.rs

```diff
@@ -425,10 +757,58 @@ impl ValidationState {
             .process_results(|iter| iter.collect())
     }
 
+    /// Checks if this edge represents a top-level mutation field.
+    fn field_if_top_level_mutation(
+        supergraph_path: &TransitionGraphPath,
+        edge_weight: &crate::query_graph::QueryGraphEdge,
+    ) -> Result<Option<FieldDefinitionPosition>, FederationError> {
+        // Must be at root level (size 0)
+        if supergraph_path.iter().count() != 0 {
+            return Ok(None);
+        }
+
+        // Must be a field collection
+        let QueryGraphEdgeTransition::FieldCollection {
+            field_definition_position,
+            ..
+        } = &edge_weight.transition
+        else {
+            return Ok(None);
+        };
+
+        // Must be from mutation root
+        let root_node = supergraph_path.head_node()?;
+        if root_node.root_kind != Some(SchemaRootDefinitionKind::Mutation) {
+            return Ok(None);
+        }
+
+        Ok(Some(field_definition_position.clone()))
+    }
+
+    /// Gets the subgraph name for a path that represents a top-level mutation field.
+    fn subgraph_of_top_level_mutation(
+        subgraph_path_info: &SubgraphPathInfo,
+    ) -> Result<Arc<str>, FederationError> {
+        let Some((last_edge, _, _, _, _)) = subgraph_path_info.path.path.iter().last() else {
+            return Err(FederationError::internal("Path unexpectedly missing edge"));
+        };
+        let (_, head_node) = subgraph_path_info
+            .path
+            .path
+            .graph()
+            .edge_endpoints(last_edge)?;
+        let head_weight = subgraph_path_info
+            .path
+            .path
+            .graph()
+            .node_weight(head_node)?;
+        Ok(head_weight.source.clone())
+    }
+
     pub(super) fn current_subgraph_context_keys(
         &self,
     ) -> Result<IndexSet<SubgraphContextKey>, FederationError> {
-        self.subgraph_paths
+        self.subgraph_path_infos
```

Comment:
The JS PR added a new parameter `subgraphPathInfos` here, which is missing here.